### PR TITLE
KAFKA-6772: Load credentials from ZK before accepting connections

### PR DESCRIPTION
**Pull request by rajinisivaram**
Start processing client connections only after completing KafkaServer initialization to ensure that credentials are loaded from ZK into cache before authentications are processed. Acceptors are started earlier so that bound port is known for registering in ZK.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by omkreddy** [comment]
LGTM.

Just for my understanding, Will there be any issue, if server starts receiving incoming connections only after completing  the initialization of the server?

**Comment by rajinisivaram** [comment]
@omkreddy Thanks for the review. At the moment we create and start the acceptors and processors in SocketServer together. We need to start acceptors early on in order to register brokers in ZK with the bound port for the listeners. My second choice was to delay starting processors, but since the simple reordering worked, I didn't actually try that out.

**Comment by rajinisivaram** [comment]
@omkreddy I tried out delayed processor start and that wasn't as big a change as I expected. So I have updated the PR. I think this will be easier to maintain compared to managing the ordering within KafkaServer.

File: core/src/main/scala/kafka/network/SocketServer.scala [hunk 1]
```diff
@@ -307,13 +316,25 @@ private[kafka] class Acceptor(val endPoint: EndPoint,
   private val nioSelector = NSelector.open()
   val serverChannel = openServerSocket(endPoint.host, endPoint.port)
   private val processors = new ArrayBuffer[Processor]()
+  private val processorsStarted = new AtomicBoolean
 
   private[network] def addProcessors(newProcessors: Buffer[Processor]): Unit = synchronized {
-    newProcessors.foreach { processor =>
+    processors ++= newProcessors
```
Comment (junrao) [inline core/src/main/scala/kafka/network/SocketServer.scala:336]: This is not directly related to this patch. However, it seems that we need some synchronization on processor between the reader and the writer?

**Review by junrao — APPROVED** [review]
@rajinisivaram : Thanks for the patch. LGTM. Just a minor comment below.

File: core/src/main/scala/kafka/network/SocketServer.scala (hunk 1 above)
Comment (rajinisivaram) [reply core/src/main/scala/kafka/network/SocketServer.scala:336]: @junrao Thanks for the review. We were already synchronizing on all read/write accesses to `processors` apart from `shutdown()`. I think that was safe anyway, but I have added synchronize in shutdown as well to be consistent.

File: core/src/main/scala/kafka/network/SocketServer.scala [hunk 2]
```diff
@@ -110,6 +113,11 @@ class SocketServer(val config: KafkaConfig, val metrics: Metrics, val time: Time
     info("Started " + acceptors.size + " acceptor threads")
   }
 
+  def startProcessors(): Unit = synchronized {
+    acceptors.values.asScala.foreach { _.startProcessors() }
+    info("Started processors for " + acceptors.size + " acceptors")
```
Comment (ijuma) [inline core/src/main/scala/kafka/network/SocketServer.scala]: Nit: use string interpolation?

File: core/src/main/scala/kafka/server/KafkaServer.scala [hunk 3]
```diff
@@ -243,8 +243,11 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
         tokenCache = new DelegationTokenCache(ScramMechanism.mechanismNames)
         credentialProvider = new CredentialProvider(ScramMechanism.mechanismNames, tokenCache)
 
+        // Create and start the socket server acceptor threads so that the bound port is known.
+        // Delay starting processors until the end of the initialization sequence to ensure
+        // that credentials have been loaded before processing authentications.
         socketServer = new SocketServer(config, metrics, time, credentialProvider)
-        socketServer.startup()
+        socketServer.startup(startupProcessors = false)
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaServer.scala:250]: We should document the new parameter in the scaladoc for `startup` and the reason why it's needed (i.e. people shouldn't have to read the comments in a particular usage to understand it). We should also include the actual behaviour impact (e.g. what happens to the requests that arrive while the processors are not started).

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, left a couple of minor comments.

**Comment by rajinisivaram** [comment]
@ijuma Thanks for the review, I have addressed the comments.

**Comment by rajinisivaram** [comment]
Merging to trunk, 1.1 and 1.0.
